### DISCUSSION TITLE
Attaching screenshots to an allure report in azure devops

### DISCUSSION THREAD

**Discussion by akotrulev** [allure-framework/allure2#2726] [Questions & Support]
Hello, I've added some code to attach a screenshot to an allure report. Running it locally works as expected - image is visible in preview, can be downloaded etc. The same code running on devops in a windows machine and using michaelclay allure viewer task. As far as I saw from the discussion there, it's a known issue but not answer from the dev. 

Is there another view task or a way to display the allure report in azure devops?

**Comment by ciconq** [comment]
Hey @akotrulev, which `task` in `Azure DevOps` are you using to run your tests? 

I had similar problems when using `VSTest@2` task to run my tests but I have found a workaround for this by specifying the custom batch size option.

Also you can try to use the official `Azure DevOps extension` which is published by `Qameta Software`.

**Reply by akotrulev** [reply]
I'm using VSTest@2 to run the tests and then using cmd, I run allure to generate the report. At the end I just need to display it in a window in the pipeline for which I use MichaelClay.azure-pipelines-allure-report-viewer.PublishAllureReport.PublishAllureReport@1 (https://marketplace.visualstudio.com/items?itemName=MichaelClay.azure-pipelines-allure-report-viewer&ssr=false#qna) and this is the problematic step. 

Looking at the report, locally the image is hosted at localhost (
http://192.168.68.65:53854/data/attachments/b778a83121bcd043.png), while in the pipeline the path is https://michaelclay.gallerycdn.vsassets.io/extensions/michaelclay/azure-pipelines-allure-report-viewer/1.0.21/1677573881498/dist/data/attachments/49437509b2fcb4fe.png which results in a BlobNotFound error.

**Reply by ciconq** [reply]
Please try using the official extension.
The one you are referring to has not been updated for a year and a half.

**Reply by ArinaNichko** [reply]
@ciconq I use PublishAllureReport@1 official extension and have the same issue with screenshots

**Reply by akotrulev** [reply]
> @ciconq I use PublishAllureReport@1 official extension and have the same issue with screenshots

I solved the issue by using https://marketplace.visualstudio.com/items?itemName=qameta.allure-azure-pipelines . Now screenshots are being attached and viewable in the report (use img/* as the format so you can preview it without downloading)

**Comment by akotrulev** [comment]
SOLUTION: 
Use https://marketplace.visualstudio.com/items?itemName=qameta.allure-azure-pipelines 
Pass the path to where all of the json files are from allure

Here is the hook for attaching screenshots

`  [AfterStep]
  public void InsertReportingSteps(ScenarioContext sc)
  {
      if (sc.TestError != null)
      {
          byte[] content = CaptureScreenshot(((DriverAdapter)sc["driver"]).NativeDriver);
          AllureLifecycle.Instance.AddAttachment("Failed screenshot", "image/*", content);
      }
  }
public byte[] CaptureScreenshot(IWebDriver driver)
{
    return ((ITakesScreenshot)driver).GetScreenshot().AsByteArray;
}
`

**Reply by ArinaNichko** [reply]
@akotrulev, I use that (https://marketplace.visualstudio.com/items?itemName=qameta.allure-azure-pipelines) and locally I have screenshots. Also I checked files in allure-results in pipeline and see here and it didn't display in report

Copying /home/vsts/work/1/s/allure-results/aed6219c-cb49-4d6b-8911-47df19f02946-attachment.png to /home/vsts/work/1/a/allure-results/aed6219c-cb49-4d6b-8911-47df19f02946-attachment.png

**Reply by akotrulev** [reply]
Can you show me the vs test task as well as the allure report? Judging by the paths (/s and /a) seems like it's in different directories (working vs current)

**Reply by ArinaNichko** [reply]
Test task: 
  - task: Maven@4
        displayName: 'Run Parallel tests'
        inputs:
          mavenPomFile: 'pom.xml'
          goals: 'clean install test -DsuiteXmlFile=src/test/resources/runners/${{parameters.runner}} 
          -Dbrowser=${{parameters.browser}} 
          -Denv=${{parameters.env}} 
          -DthreadCount=${{parameters.threadCount}} 
          -Dbrowserstack=${{parameters.browserstackRun}}'
          javaHomeOption: 'JDKVersion'
          jdkVersionOption: '1.17'
          mavenVersionOption: 'Default'
          mavenAuthenticateFeed: false
          effectivePomSkip: false
          sonarQubeRunAnalysis: false

      - task: Maven@4
        displayName: 'Run Not-Parallel tests'
        inputs:
          mavenPomFile: 'pom.xml'
          goals: 'clean install test -DsuiteXmlFile=src/test/resources/runners/testngOryna.xml
         -Dbrowser=${{parameters.browser}}
         -Denv=${{parameters.env}}
         -DthreadCount=1
         -Dbrowserstack=${{parameters.browserstackRun}}'
          javaHomeOption: 'JDKVersion'
          jdkVersionOption: '1.17'
          mavenVersionOption: 'Default'
          mavenAuthenticateFeed: false
          effectivePomSkip: false
          sonarQubeRunAnalysis: false
          
            - task: PublishAllureReport@1
        displayName: 'Publish Allure Report'
        inputs:
          allureVersion: 2.27.0
          testResultsDir: allure-results
        condition: always()

      - task: CopyFiles@2
        inputs:
          SourceFolder: '$(System.DefaultWorkingDirectory)/allure-results'
          Contents: '**'  # Matches all files and subdirectories
          TargetFolder: '$(Build.artifactstagingdirectory)/allure-results'
          cleanTargetFolder: true
        condition: always()
        
        found 19 files
Cleaning target folder: /home/vsts/work/1/a/allure-results
Copying /home/vsts/work/1/s/allure-results/020ee996-7770-48db-879b-3e70a1f66506-container.json to /home/vsts/work/1/a/allure-results/020ee996-7770-48db-879b-3e70a1f66506-container.json
Copying /home/vsts/work/1/s/allure-results/0d199333-ea5d-4f55-9cbb-e940077d7c75-container.json to /home/vsts/work/1/a/allure-results/0d199333-ea5d-4f55-9cbb-e940077d7c75-container.json
Copying /home/vsts/work/1/s/allure-results/1a3571b8-9340-4b04-84d3-4cb5539601f1-container.json to /home/vsts/work/1/a/allure-results/1a3571b8-9340-4b04-84d3-4cb5539601f1-container.json
Copying /home/vsts/work/1/s/allure-results/239d9425-03f5-490e-a414-3c5d14b2a4b5-container.json to /home/vsts/work/1/a/allure-results/239d9425-03f5-490e-a414-3c5d14b2a4b5-container.json
Copying /home/vsts/work/1/s/allure-results/39248a0a-2cfb-48a8-bbbd-5962e5b8889e-container.json to /home/vsts/work/1/a/allure-results/39248a0a-2cfb-48a8-bbbd-5962e5b8889e-container.json
Copying /home/vsts/work/1/s/allure-results/52351d66-4ced-4f27-966f-b74ce70faa76-container.json to /home/vsts/work/1/a/allure-results/52351d66-4ced-4f27-966f-b74ce70faa76-container.json
Copying /home/vsts/work/1/s/allure-results/57b28ba4-4ec4-4094-baf2-7df970aee358-container.json to /home/vsts/work/1/a/allure-results/57b28ba4-4ec4-4094-baf2-7df970aee358-container.json
Copying /home/vsts/work/1/s/allure-results/5f976b74-89ed-4416-bcd1-54992d6c6a7b-container.json to /home/vsts/work/1/a/allure-results/5f976b74-89ed-4416-bcd1-54992d6c6a7b-container.json
Copying /home/vsts/work/1/s/allure-results/752bf617-f26c-4021-a2ed-1c761b67a371-container.json to /home/vsts/work/1/a/allure-results/752bf617-f26c-4021-a2ed-1c761b67a371-container.json
Copying /home/vsts/work/1/s/allure-results/763d0b7f-eba8-4008-8c4d-21ed7b6982f1-result.json to /home/vsts/work/1/a/allure-results/763d0b7f-eba8-4008-8c4d-21ed7b6982f1-result.json
Copying /home/vsts/work/1/s/allure-results/82da1a46-f6c0-4c85-9bba-d5add5f2d418-container.json to /home/vsts/work/1/a/allure-results/82da1a46-f6c0-4c85-9bba-d5add5f2d418-container.json
Copying /home/vsts/work/1/s/allure-results/870588d1-a607-4572-a68a-c0f1375bad0f-result.json to /home/vsts/work/1/a/allure-results/870588d1-a607-4572-a68a-c0f1375bad0f-result.json
Copying /home/vsts/work/1/s/allure-results/8a27aed8-043b-44bd-ba47-656e55cccf53-container.json to /home/vsts/work/1/a/allure-results/8a27aed8-043b-44bd-ba47-656e55cccf53-container.json
Copying /home/vsts/work/1/s/allure-results/9ff268ae-3b24-4a0a-b6c7-09afd7002337-container.json to /home/vsts/work/1/a/allure-results/9ff268ae-3b24-4a0a-b6c7-09afd7002337-container.json
Copying /home/vsts/work/1/s/allure-results/aed6219c-cb49-4d6b-8911-47df19f02946-attachment.png to /home/vsts/work/1/a/allure-results/aed6219c-cb49-4d6b-8911-47df19f02946-attachment.png
Copying /home/vsts/work/1/s/allure-results/bd5d20f6-fa8f-47a0-9369-cced4d33bcec-result.json to /home/vsts/work/1/a/allure-results/bd5d20f6-fa8f-47a0-9369-cced4d33bcec-result.json
Copying /home/vsts/work/1/s/allure-results/ca1fb62a-a8d7-45bd-b2ba-11bd8b24637b-container.json to /home/vsts/work/1/a/allure-results/ca1fb62a-a8d7-45bd-b2ba-11bd8b24637b-container.json
Copying /home/vsts/work/1/s/allure-results/cbf33121-976d-4a38-893f-eda60edb959a-container.json to /home/vsts/work/1/a/allure-results/cbf33121-976d-4a38-893f-eda60edb959a-container.json
Copying /home/vsts/work/1/s/allure-results/fcb8c835-80fb-41cf-a34b-b9055756971e-container.json to /home/vsts/work/1/a/allure-results/fcb8c835-80fb-41cf-a34b-b9055756971e-container.json
Finishing: CopyFiles

**Reply by akotrulev** [reply]
If you don't want to have the history, i highly recommend deleting the folder as one of the first steps of the pipeline 
'
del /f /s /q  $(System.DefaultWorkingDirectory)\allure-results 1>nul
rmdir /s /q  $(System.DefaultWorkingDirectory)\allure-results
' 

Also in the publish allure result task, provide the full path to the .json files (looking at your tasks, it should be $(System.DefaultWorkingDirectory)/allure-results

Also you can try by removing CopyFiles@2 task and see if that helps at all

**Reply by ArinaNichko** [reply]
As I understand, it sees my json files it generates report, but without screenshots, but when I checked files in allure-results folder I have image

**Reply by akotrulev** [reply]
From the logs it looks like it copies only the json files, can you try to build the report from the original directory?

**Reply by ArinaNichko** [reply]
I could build from original, but pay attention on list of coped files I have one png, so it attaches to the folder where our report is generated 

found 19 files
Cleaning target folder: /home/vsts/work/1/a/allure-results
Copying /home/vsts/work/1/s/allure-results/020ee996-7770-48db-879b-3e70a1f66506-container.json to /home/vsts/work/1/a/allure-results/020ee996-7770-48db-879b-3e70a1f66506-container.json
Copying /home/vsts/work/1/s/allure-results/0d199333-ea5d-4f55-9cbb-e940077d7c75-container.json to /home/vsts/work/1/a/allure-results/0d199333-ea5d-4f55-9cbb-e940077d7c75-container.json
Copying /home/vsts/work/1/s/allure-results/1a3571b8-9340-4b04-84d3-4cb5539601f1-container.json to /home/vsts/work/1/a/allure-results/1a3571b8-9340-4b04-84d3-4cb5539601f1-container.json
Copying /home/vsts/work/1/s/allure-results/239d9425-03f5-490e-a414-3c5d14b2a4b5-container.json to /home/vsts/work/1/a/allure-results/239d9425-03f5-490e-a414-3c5d14b2a4b5-container.json
Copying /home/vsts/work/1/s/allure-results/39248a0a-2cfb-48a8-bbbd-5962e5b8889e-container.json to /home/vsts/work/1/a/allure-results/39248a0a-2cfb-48a8-bbbd-5962e5b8889e-container.json
Copying /home/vsts/work/1/s/allure-results/52351d66-4ced-4f27-966f-b74ce70faa76-container.json to /home/vsts/work/1/a/allure-results/52351d66-4ced-4f27-966f-b74ce70faa76-container.json
Copying /home/vsts/work/1/s/allure-results/57b28ba4-4ec4-4094-baf2-7df970aee358-container.json to /home/vsts/work/1/a/allure-results/57b28ba4-4ec4-4094-baf2-7df970aee358-container.json
Copying /home/vsts/work/1/s/allure-results/5f976b74-89ed-4416-bcd1-54992d6c6a7b-container.json to /home/vsts/work/1/a/allure-results/5f976b74-89ed-4416-bcd1-54992d6c6a7b-container.json
Copying /home/vsts/work/1/s/allure-results/752bf617-f26c-4021-a2ed-1c761b67a371-container.json to /home/vsts/work/1/a/allure-results/752bf617-f26c-4021-a2ed-1c761b67a371-container.json
Copying /home/vsts/work/1/s/allure-results/763d0b7f-eba8-4008-8c4d-21ed7b6982f1-result.json to /home/vsts/work/1/a/allure-results/763d0b7f-eba8-4008-8c4d-21ed7b6982f1-result.json
Copying /home/vsts/work/1/s/allure-results/82da1a46-f6c0-4c85-9bba-d5add5f2d418-container.json to /home/vsts/work/1/a/allure-results/82da1a46-f6c0-4c85-9bba-d5add5f2d418-container.json
Copying /home/vsts/work/1/s/allure-results/870588d1-a607-4572-a68a-c0f1375bad0f-result.json to /home/vsts/work/1/a/allure-results/870588d1-a607-4572-a68a-c0f1375bad0f-result.json
Copying /home/vsts/work/1/s/allure-results/8a27aed8-043b-44bd-ba47-656e55cccf53-container.json to /home/vsts/work/1/a/allure-results/8a27aed8-043b-44bd-ba47-656e55cccf53-container.json
Copying /home/vsts/work/1/s/allure-results/9ff268ae-3b24-4a0a-b6c7-09afd7002337-container.json to /home/vsts/work/1/a/allure-results/9ff268ae-3b24-4a0a-b6c7-09afd7002337-container.json
### Copying /home/vsts/work/1/s/allure-results/aed6219c-cb49-4d6b-8911-47df19f02946-attachment.png to /home/vsts/work/1/a/allure-results/aed6219c-cb49-4d6b-8911-47df19f02946-attachment.png
Copying /home/vsts/work/1/s/allure-results/bd5d20f6-fa8f-47a0-9369-cced4d33bcec-result.json to /home/vsts/work/1/a/allure-results/bd5d20f6-fa8f-47a0-9369-cced4d33bcec-result.json
Copying /home/vsts/work/1/s/allure-results/ca1fb62a-a8d7-45bd-b2ba-11bd8b24637b-container.json to /home/vsts/work/1/a/allure-results/ca1fb62a-a8d7-45bd-b2ba-11bd8b24637b-container.json
Copying /home/vsts/work/1/s/allure-results/cbf33121-976d-4a38-893f-eda60edb959a-container.json to /home/vsts/work/1/a/allure-results/cbf33121-976d-4a38-893f-eda60edb959a-container.json
Copying /home/vsts/work/1/s/allure-results/fcb8c835-80fb-41cf-a34b-b9055756971e-container.json to /home/vsts/work/1/a/allure-results/fcb8c835-80fb-41cf-a34b-b9055756971e-container.json
Finishing: CopyFiles

**Reply by eugenegoncharuk** [reply]
hi. @ArinaNichko Have you fixed the issue with images not shown?
